### PR TITLE
[Tabular][FastAI] Upgrade fastai v1 -> fastai2

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*
Re-write of `NeuralNetFastAI` from fast.ai v1 to v2

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-977-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/2/index.html

File: tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py [hunk 1]
```diff
@@ -405,3 +400,37 @@ def _get_default_auxiliary_params(self) -> dict:
...
+
+    def __get_metrics_map(self):
+        from fastai.metrics import rmse, mse, mae, accuracy, FBeta, RocAucBinary, Precision, Recall, R2Score
+        metrics_map = {
+            # Regression
+            'root_mean_squared_error': rmse,
+            'mean_squared_error': mse,
+            'mean_absolute_error': mae,
+            'r2': R2Score(),
+            # Not supported: median_absolute_error
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:412]: since eval_metric is mainly just used for early-stopping, maybe if a metric is not supported, you can still map it to the most similar metric for training the FastAI net?  Eg. for median_absolute_error, you could choose mae. And you could print low-priority output debug message that:  training objective for FastAI net is actually mean_absolute_error instead of median_absolute_error because the latter is not supported as a training objective.

**Review by jwmueller — APPROVED** [review]
Looks good to me.  You'll first run benchmark to check for accuracy/efficiency/memory degradation before merging?

File: tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py (hunk 1 above)
Comment (gradientsky) [reply tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:387]: I ended up implementing it myself - update PR.

**Comment by gradientsky** [comment]
> Looks good to me. You'll first run benchmark to check for accuracy/efficiency/memory degradation before merging?

Yes, this needs to be done before the merge.

**Comment by szha** [comment]
Job PR-977-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/3/index.html

**Comment by szha** [comment]
Job PR-977-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/5/index.html

**Comment by szha** [comment]
Job PR-977-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/9/index.html

**Comment by jwmueller** [comment]
@gradientsky Any updates on benchmarking/merging this? 
Taesup is hoping to adapt the fastAI net for our quantile-regression task, since his existing NN-quantile code is also in pytorch.

**Comment by Innixma** [comment]
re benchmarking, I'm still waiting for automlbenchmark to merge my PR to enable our team to easily benchmark AG: https://github.com/openml/automlbenchmark/pull/237

**Comment by szha** [comment]
Job PR-977-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/11/index.html

**Comment by szha** [comment]
Job PR-977-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/12/index.html

**Comment by szha** [comment]
Job PR-977-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/13/index.html

**Comment by szha** [comment]
Job PR-977-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/15/index.html

**Review by Innixma — APPROVED** [review]
Benchmarks look good: 2x training speedup, 4x inference speedup, roughly equal model quality.

File: tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py [hunk 2]
```diff
@@ -222,123 +219,88 @@ def _fit(self,
...
             self.params_trained['best_epoch'] = save_callback.best_epoch
 
     def _generate_datasets(self, X, y, X_val, y_val):
         df_train = pd.concat([X, X_val], ignore_index=True)
         df_train[LABEL] = pd.concat([y, y_val], ignore_index=True)
         train_idx = np.arange(len(X))
         if X_val is None:
-            val_idx = train_idx  # use validation set for refit_full case - it's not going to be used for early stopping
+            val_idx = train_idx + len(train_idx)  # use validation set for refit_full case - it's not going to be used for early stopping
+            df_train = pd.concat([df_train, df_train], ignore_index=True)
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply tabular/src/autogluon/tabular/models/fastainn/tabular_nn_fastai.py:272]: doesn't work - fastai is inflexible here: what is included in validation set is excluded from training set i.e. if row 1 is in val_idx, it won't be in train_idx.

**Comment by szha** [comment]
Job PR-977-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-977/16/index.html
